hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/batching/PartitionAwareCallableBatchingRunnable.java

```diff
@@ -0,0 +1,128 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.spi.impl.operationservice.impl.batching;
+
+import com.hazelcast.core.ICompletableFuture;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.nio.Address;
+import com.hazelcast.spi.NodeEngine;
+import com.hazelcast.spi.impl.AbstractCompletableFuture;
+import com.hazelcast.spi.impl.operationexecutor.impl.PartitionOperationThread;
+import com.hazelcast.spi.impl.operationservice.impl.OperationServiceImpl;
+import com.hazelcast.spi.partition.IPartition;
+import com.hazelcast.spi.partition.IPartitionService;
+
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * If run on partition thread, it will discover what thread it is run on, and it will run an instance of
+ * PartitionAwareCallable on all partitions this thread is responsible for.
+ *
+ * PartitionAwareCallable will be created using the PartitionAwareCallableFactory class.
+ * THere will be one PartitionAwareCallable per all partitions the runnable will run the processing for.
+ *
+ * IMPORTANT:
+ * <ul>
+ * <li>Runnable assumes to be run on a partition-thread. If not run on a partition-thread a RuntimeException will be thrown.</li>
+ * <li>This optimization highly relies on the threading-model and has been implemented to speed up invocations that span all
+ * partitions, e.g. HD-index-query, or potentially many more like IMap.size(), etc.</li>
+ * </ul>
+ *
+ * @see PartitionAwareCallable
+ * @see PartitionAwareCallableFactory
+ * @since 3.9
+ */
+public class PartitionAwareCallableBatchingRunnable implements Runnable {
+
+    private final PartitionAwareCallableFactory factory;
+    private final IPartitionService partitionService;
+    private final Address memberAddress;
+    private final int partitionThreadCount;
+
+    private final CopyOnWriteArrayList results = new CopyOnWriteArrayList();
+    private final AtomicInteger finished = new AtomicInteger(0);
+
+    private final ResultFuture future;
+
+    /**
+     * @param nodeEngine nodeEngine
+     * @param factory    factory of PartitionAwareCallable to be run on the partition thread.
+     */
+    public PartitionAwareCallableBatchingRunnable(NodeEngine nodeEngine, PartitionAwareCallableFactory factory) {
+        this.factory = factory;
+        this.partitionService = nodeEngine.getPartitionService();
+        this.memberAddress = nodeEngine.getThisAddress();
+        this.partitionThreadCount = ((OperationServiceImpl) nodeEngine.getOperationService()).getPartitionThreadCount();
+        this.future = new ResultFuture(nodeEngine, nodeEngine.getLogger(getClass()));
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public void run() {
+        if (!(Thread.currentThread() instanceof PartitionOperationThread)) {
```

Comment:
do we need to check this every time, any possibility to escape out of partition threads in production? seems this can be replaceable with `assert Thread.currentThread() instanceof PartitionOperationThread`

Reply:
Also, we have `ThreadUtil#assertRunningOnPartitionThread`

Reply:
Fixed.